sql/core/src/test/scala/org/apache/spark/sql/CollationSQLExpressionsSuite.scala

```diff
@@ -2435,829 +2448,878 @@ class CollationSQLExpressionsSuite
   }
 
   test("min_by supports collation") {
-    val collation = "UNICODE"
-    val query = s"SELECT min_by(x, y) FROM VALUES ('a', 10), ('b', 50), ('c', 20) AS tab(x, y);"
-    withSQLConf(SqlApiConf.DEFAULT_COLLATION -> collation) {
-      checkAnswer(
-        sql(query),
-        Seq(
-          Row("a")
+    testAdditionalCollations.foreach(collation => {
+      val query = s"SELECT min_by(x, y) FROM VALUES ('a', 10), ('b', 50), ('c', 20) AS tab(x, y);"
+      withSQLConf(SqlApiConf.DEFAULT_COLLATION -> collation) {
+        checkAnswer(
+          sql(query),
+          Seq(
+            Row("a")
+          )
         )
-      )
-      // check result row data type
-      val dataType = StringType(collation)
-      assert(sql(query).schema.head.dataType == dataType)
-    }
+        // check result row data type
+        val dataType = StringType(collation)
+        assert(sql(query).schema.head.dataType == dataType)
+      }
+    })
   }
 
   test("max_by supports collation") {
-    val collation = "UNICODE"
-    val query = s"SELECT max_by(x, y) FROM VALUES ('a', 10), ('b', 50), ('c', 20) AS tab(x, y);"
-    withSQLConf(SqlApiConf.DEFAULT_COLLATION -> collation) {
-      checkAnswer(
-        sql(query),
-        Seq(
-          Row("b")
+    testAdditionalCollations.foreach(collation => {
+      val query = s"SELECT max_by(x, y) FROM VALUES ('a', 10), ('b', 50), ('c', 20) AS tab(x, y);"
+      withSQLConf(SqlApiConf.DEFAULT_COLLATION -> collation) {
+        checkAnswer(
+          sql(query),
+          Seq(
+            Row("b")
+          )
         )
-      )
-      // check result row data type
-      val dataType = StringType(collation)
-      assert(sql(query).schema.head.dataType == dataType)
-    }
+        // check result row data type
+        val dataType = StringType(collation)
+        assert(sql(query).schema.head.dataType == dataType)
+      }
+    })
   }
 
   test("array supports collation") {
-    val collation = "UNICODE"
-    val query = s"SELECT array('a', 'b', 'c');"
-    withSQLConf(SqlApiConf.DEFAULT_COLLATION -> collation) {
-      checkAnswer(
-        sql(query),
-        Seq(
-          Row(Seq("a", "b", "c"))
+    testAdditionalCollations.foreach(collation => {
+      val query = s"SELECT array('a', 'b', 'c');"
+      withSQLConf(SqlApiConf.DEFAULT_COLLATION -> collation) {
+        checkAnswer(
+          sql(query),
+          Seq(
+            Row(Seq("a", "b", "c"))
+          )
         )
-      )
-      // check result row data type
-      val dataType = ArrayType(StringType(collation), false)
-      assert(sql(query).schema.head.dataType == dataType)
-    }
+        // check result row data type
+        val dataType = ArrayType(StringType(collation), false)
+        assert(sql(query).schema.head.dataType == dataType)
+      }
+    })
   }
 
   test("array_agg supports collation") {
```

Comment:
Those all tests are similar. Let's rewrite in the way:
```scala
    testAdditionalCollations.foreach { collation =>
      Seq(
        "SELECT array('a', 'b', 'c')" -> Seq(Row(Seq("a", "b", "c"))),
        "SELECT array_agg(col) FROM VALUES ('a'), ('b'), ('c') AS tab(col)" ->
          Seq(Row(Seq("a", "b", "c")))
      ).foreach { case (query, result) =>
        withSQLConf(SqlApiConf.DEFAULT_COLLATION -> collation) {
          checkAnswer(sql(query), result)
          // check result row data type
          val dataType = ArrayType(StringType(collation), false)
          assert(sql(query).schema.head.dataType == dataType)
        }
      }
    }
```

Reply:
I made a slightly different refactoring, still removing code duplication across all similar tests as you pointed out. The method testCollationSqlExpressionCommon() now does the common part, whereas I preferred to keep all the tests separately as troubleshooting is easier that way (e.g. if a specific test fails, test name immediately indicates which one it is) as opposed to putting all the test cases under a single test.

Please let me know if you have any additional suggestions/concerns, thnx!

Reply:
Also, to clarify why I didn't apply the refactoring in the exact way that you suggested above - dataType value is defined differently throughout various tests, sometimes in a bit more complex manner, so it wouldn't be practical to include it in the sequence as well, i.e. I felt that it would reduce readability of the tests/coverage.
Hence, I've kept each test separately, where query, result and dataType are explicitly specified, whereas everything else is moved to the common method testCollationSqlExpressionCommon()

Reply:
@MaxGekk , please take another look at the changes and let me know if you see further issues, or the PR could be merged, thanks!